rfc/rfc-2/2 - Deletion Info Conversion.md

```diff
@@ -0,0 +1,159 @@
+```
+
+If a file with an existing deletion vector is updated again, a new deletion vector is created, merging the previous and 
+new deletions. The commit log records the removal of the old deletion vector information and the addition of the new 
+one.
+
+Example:
+```
+Type       Path           DV Path
+RemoveFile file_a.parquet file_a_dv_1.bin
+AddFile    file_a.parquet file_a_dv_2.bin
+```
+
+When a compaction operation rewrites records into new files, and deletion vectors associated with data files that are
+removed are also marked as deleted. The commit log records the removal of the old file and deletion vector and 
+the addition of the new compacted file.
+
+Example:
+```
+Type       Path           DV Path
+RemoveFile file_a.parquet file_a_dv_2.bin
+AddFile    file_d.parquet NULL
+```
+
+## Implementation of the conversion of Deletion Vectors in XTable
+On a high level, the implementation of deletion vector conversion in XTable involves the following steps. XTable must 
+first locate the deletion vector, which can be stored either inline within the commit logs or can be stored in a 
+separate deletion vector file whose path is recorded in the commit logs. Once located, XTable can utilize the utility 
+methods provided by table format libraries to stream the ordinals of the records affected by the deletion operations.
+
+Unlike currently supported conversion of metadata in XTable, deletion vectors require special handling. While all
+table metadata is present in commit logs and manifest files, deletion vectors can be stored in separate files. The size
+of the deletion vectors is proportional to the number of records deleted, which can be significant in large tables.
+The proposed implementation aims to handle this efficiently by streaming the records from the sources to the targets.
+
+**New Class: `InternalDeletionVector`**
+To improve the functionality, a new class to manage the internal in-memory representation of deletion vectors is added.
+This is similar to creating internal representation for data files, schema, and partitioning spec.
+
+The following metadata is relevant for deletion vectors:
+- `dataFilePath`: Path of the data file associated with the deletion vector.
+- `size`: Size of the deletion vector.
+- `countRecordsDeleted`: Number of records deleted.
+- `sourceDeletionVectorFilePath`: Path of the deletion vector file (optional as deletion vectors can be stored inline).
+- `offset`: Offset of the deletion vector within the file (optional as inline deletion vectors do not have an offset).
+- `binaryRepresentation`: Binary representation of the inline deletion vector (optional).
+- `ordinalStream`: Stream of ordinals of the records deleted.
+
+**Modifications in `TableChange`**
+Currently XTable only captures the addition and removal of data files. To support deletion vectors, the `TableChange`
+class is updated to include the following fields a new field `deletionVectorsAdded` to capture deletion vectors added 
```

Comment:
I didn't understand how will deletion vectors removed (or merged with existing ones) will be synced in XTable ? Asking this assuming this is behavior for deletion vectors behavior.
```
If a file with an existing deletion vector is updated again, a new deletion vector is created, merging the previous and 
new deletions. The commit log records the removal of the old deletion vector information and the addition of the new one.
```

Reply:
I have add code changes specific to Delta (pull request #627) and Iceberg [ work-in-progress branch (url)](https://github.com/apache/incubator-xtable/commit/eb6a8c7b363d46d33bb82aa932aea18d12fe8f89#diff-b700eabfdce8dd7c376e280789e6be74d550c5b4a9be76de45185c7c73f94e9d)

It might be challenging to cover the delta protocol details comprehensively in this RFC. I'm happy to schedule a call to clarify your questions and provide more details if the code changes are not sufficient. What do you think?